### PR TITLE
Implement basic logic of re-using buffers

### LINKED ISSUE #353: Very high CPU utilization when using native driver
### Environment

- OS version: Centos 7.5
- JDK version: AdoptOpenJDK 11.0.7
- ClickHouse Server version: `21.5.5 revision 54448`
- ClickHouse Native JDBC version: `2.5.6`
- (Optional) Spark version: N/A
- (Optional) Other components' version: N/A
- Kafkaconnect version: `Confluent distribution 6.1.1-2.12`
- Postgresql version: `12.3`

### Error logs
N/A

### Steps to reproduce

1. Heavy load topic (125K-200K messages per second) with 50 partitions
2. 3 node kafkaconnect cluster (physical hosts, not VM) with 16 vCPU (8c/16t) each, 128 GB RAM
3. Clickhouse standalone server (RAID10 on SATA3 SSDs)
4. Postgresql standalone server (RAID10 on SATA3 SSDs)
5. 2 JDBC sink connectors to ingest data, one to Postgres and one to Clickhouse using this JDBC driver

### Other descriptions
With setup above, when we run Postgres ingest the CPU utilization across all 3 KC nodes is very light, about 5% (loadavg5 about 1.2 on each kafkaconnect server). When we stop Postgres ingest and start Clickhouse ingest CPU utilization goes to 100% across all 3 nodes (loadavg5 about 18-20 on each kafkaconnect server).

There seems to be some code that causes very high CPU utilization when using this driver which exceeds any load expectations. As compared to Postgresql JDBC driver this one uses about 15-20 times more CPU resources.

### PR DESCRIPTION
**Pull request by akamensky**
Closes #353

### CODE REVIEW

File: clickhouse-native-jdbc/src/main/java/com/github/housepower/serde/BinarySerializer.java [hunk 1]
```diff
@@ -27,14 +27,17 @@
...
+    
     public BinarySerializer(BuffedWriter writer, boolean enableCompress) {
         this.enableCompress = enableCompress;
         BuffedWriter compressWriter = null;
         if (enableCompress) {
             compressWriter = new CompressedBuffedWriter(ClickHouseDefines.SOCKET_SEND_BUFFER_BYTES, writer);
         }
         switcher = new Switcher<>(compressWriter, writer);
+        // max num of byte is 8 for double and long
+        writeBuffer = new byte[8];
```
Comment (sundy-li) [inline clickhouse-native-jdbc/src/main/java/com/github/housepower/serde/BinarySerializer.java:40]: 👍

**Comment by AndreyNudko** [comment]
First of all, thanks a lot for addressing this over-allocation problem. I was contemplating a simpler change, but this one is a lot better!
I have couple of questions if that's ok:

1. Is it necessary for `SOCKET_SEND_BUFFER_BYTES`, `SOCKET_RECV_BUFFER_BYTES`, and `COLUMN_BUFFER_BYTES` to be final?
We're currently setting lower values to minimize allocation
I appreciate this PR addresses the problem in a better way, but they seem like useful tunables (unless ClickHouseDefines is considered internal API)

2. If there was a spike in writes and ByteArrayWriter allocated extra buffers - they will remain in freeList forever and won't be collected, right?
It may be desirable in case of long-term change in traffic pattern, or wasting memory in case of a short burst
Not saying this is a problem (almost certainly not for our use-cases), just trying to understand implications

3. ColumnWriterBuffer::writeTo
Is there any reason to use temporary buffer instead of doing something like 
`serializer.writeBytes(buffer.array(), buffer.arrayOffset(), buffer.remaining());`
I understand this is coupled with ByteArrayWriter implementation allocating heap buffers - but both classes already seem to have quite close relationship; and the clue is in the name

**Comment by akamensky** [comment]
@AndreyNudko Please see comment in linked ticket. I am not the right person to ask those questions and not sure the right person will have more time to work on this to address those. I am but a messenger.

**Comment by AndreyNudko** [comment]
@akamensky Don't get me wrong - those questions are more for maintainers to consider, I'm certainly not asking for more from you guys. And I really appreciate the effort you made.

**Comment by sundy-li** [comment]
@akamensky  Thanks for the pr, I'll take time to review this.

@AndreyNudko 

> Is it necessary for SOCKET_SEND_BUFFER_BYTES, SOCKET_RECV_BUFFER_BYTES, and COLUMN_BUFFER_BYTES to be final?

I do think letting it be the way before is better, users can modify set these to other values.

> If there was a spike in writes and ByteArrayWriter allocated extra buffers - they will remain in freeList forever and won't be collected, right?

Yes, that's the side effect, manually re-using the buffer or let them be controlled by JVM. But I think it may be worthy to do if we can set fixed blocksize (batchsize) for each insert( better let the value lesser than 1048576).

clickhouse-go seems to have a buffer block in each connection, see:  https://github.com/ClickHouse/clickhouse-go/blob/master/clickhouse.go#L46-L61

File: clickhouse-native-jdbc/src/main/java/com/github/housepower/data/ColumnWriterBufferFactory.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        if (pop == null) {
+            return new ColumnWriterBuffer();
+        } else {
+            pop.reset();
+            return pop;
+        }
+    }
+
+    public void recycleBuffer(ColumnWriterBuffer buffer) {
+        stack.addLast(buffer);
```
Comment (sundy-li) [inline clickhouse-native-jdbc/src/main/java/com/github/housepower/data/ColumnWriterBufferFactory.java:58]: how about reset it before `addLast`, making stack keeps all clean buffers.

File: clickhouse-native-jdbc/src/main/java/com/github/housepower/data/ColumnWriterBufferFactory.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ * Factory of column writer that can recycle {@link ColumnWriterBuffer}.
+ * 
+ * <p>
+ * The factory is thread-safe and can be used by multiple threads or connections.
+ * </p>
+ *
+ */
+public class ColumnWriterBufferFactory {
+
+    private final ConcurrentLinkedDeque<ColumnWriterBuffer> stack = new ConcurrentLinkedDeque<>();
```
Comment (sundy-li) [inline clickhouse-native-jdbc/src/main/java/com/github/housepower/data/ColumnWriterBufferFactory.java:28]: I think it's better to export a method named `clear` to clear all buffers manually.

**Comment by akamensky** [comment]
@sundy-li  as I mentioned, please check the comment in linked ticket. I did not write those changes, it comes from  a colleague who is a Java expert. They do not have more time to invest into this changeset. They also mentioned that the implementation is very crude, so you may be right on those comments you left. I am not the right person to follow up on those comments, but feel free to modify these changes as you see appropriate.

Meantime we are using those changes internally and it reduced CPU load dramatically.

**Review by sundy-li — APPROVED** [review]

**Pull request merged**